File: Medium/collect_rocks/collect_rocks.cpp
#include <iostream>
#include <vector>
using ll=long long;
using namespace std;
int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    ll m,n,k,nowr,nowc; cin>>m>>n>>k>>nowr>>nowc;
    vector<vector<ll>> grid(m,vector<ll> (n));
    for (ll i=0;i<m;i++)
        for (ll j=0;j<n;j++)
            cin>>grid[i][j];
    if (!grid[nowr][nowc])
    { cout<<0; return 0; }
    ll score=grid[nowr][nowc];
    ll rock=1;
    grid[nowr][nowc]--;
    ll nowdir=0;
    while (true)
    {
        ll newr=nowr,newc=nowc;
        if (nowdir==0) newc++;
        else if (nowdir==1) newr++;
        else if (nowdir==2) newc--;
        else newr--;
        if (newr<0 || newc<0 || newr>=m || newc>=n || grid[newr][newc]==-1)
        {
            nowdir=(nowdir+1)%4;
            continue;
        }
        nowr=newr; nowc=newc;
        if (!grid[nowr][nowc]) break;
        score+=grid[nowr][nowc];
        grid[nowr][nowc]--;
        rock++;
        if (!(score%k)) nowdir=(nowdir+1)%4;
    }
    cout<<rock;
    return 0;
}

File: Medium/collect_rocks/collect_rocks.py
import sys
input=sys.stdin.readline
m,n,k,nowr,nowc=map(int,input().split())
grid=[]
for _ in range(m):
	grid.append(list(map(int,input().split())))
if grid[nowr][nowc]==0:
	print(0)
	sys.exit(0)
score=grid[nowr][nowc]
grid[nowr][nowc]-=1
nowdir=1 if score%k==0 else 0
rock=1
while True:
	newr,newc=nowr,nowc
	if nowdir==0:
		newc+=1
	elif nowdir==2:
		newc-=1
	elif nowdir==1:
		newr+=1
	else:
		newr-=1
	if newr<0 or newc<0 or newr>=m or newc>=n or grid[newr][newc]==-1:
		nowdir=(nowdir+1)%4
		continue
	if grid[newr][newc]==0:
		break
	nowr,nowc=newr,newc
	score+=grid[nowr][nowc]
	rock+=1
	grid[nowr][nowc]-=1
	if score%k==0:
		nowdir=(nowdir+1)%4
print(rock)